### PR TITLE
Dynamic GP compatible with Tx parallel

### PR DESCRIPTION
**Pull request by LeoGuo621**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/app_abci.go [hunk 1]
```diff
@@ -26,13 +23,13 @@ func (app *OKExChainApp) DeliverTx(req abci.RequestDeliverTx) (res abci.Response
...
 	resp := app.BaseApp.DeliverTx(req)
 
-	if appconfig.GetOecConfig().GetDynamicGpMode() != types.CloseMode {
-		tx, err := evm.TxDecoder(app.marshal)(req.Tx)
-		if err == nil {
-			//optimize get tx gas price can not get value from verifySign method
-			app.gpo.CurrentBlockGPs.Update(tx.GetGasPrice(), uint64(resp.GasUsed))
-		}
-	}
+	//if appconfig.GetOecConfig().GetDynamicGpMode() != types.MinimalGpMode {
```
Comment (zjg555543) [inline app/app_abci.go]: Delete the invalid annotations?

File: app/app_abci.go [hunk 2]
```diff
@@ -46,17 +43,17 @@ func (app *OKExChainApp) DeliverRealTx(req abci.TxEssentials) (res abci.Response
...
-		tx, _ := req.(sdk.Tx)
-		if tx == nil {
-			tx, err = evm.TxDecoder(app.Codec())(req.GetRaw())
-		}
-		if err == nil {
-			//optimize get tx gas price can not get value from verifySign method
-			app.gpo.CurrentBlockGPs.Update(tx.GetGasPrice(), uint64(resp.GasUsed))
-		}
-	}
+	//var err error
```
Comment (zjg555543) [inline app/app_abci.go]: Delete the invalid annotations too

File: app/types/gasprice.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ const (
 
 	CongestionHigherGpMode = 0
 	NormalGpMode           = 1
```
Comment (zjg555543) [inline app/types/gasprice.go:13]: Does it have not used this variable NormalGpMode？

File: app/app.go [hunk 4]
```diff
@@ -646,6 +646,11 @@ func NewOKExChainApp(
 	app.SetGetTxFeeHandler(getTxFeeHandler())
 	app.SetEvmSysContractAddressHandler(NewEvmSysContractAddressHandler(app.EvmKeeper))
 	app.SetEvmWatcherCollector(app.EvmKeeper.Watcher.Collect)
+	if appconfig.GetOecConfig().GetDynamicGpMode() != types.MinimalGpMode {
```
Comment (zjg555543) [inline app/app.go]: It may be crash, when set init MinimalGpMode, but set dynamic not MinimalGpMode

File: app/app.go [hunk 5]
```diff
@@ -705,7 +706,7 @@ func (app *OKExChainApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBloc
 
 // EndBlocker updates every end block
 func (app *OKExChainApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-	if appconfig.GetOecConfig().GetDynamicGpMode() != types.CloseMode {
+	if appconfig.GetOecConfig().GetDynamicGpMode() != types.MinimalGpMode {
```
Comment (zjg555543) [inline app/app.go:708]: It may be crash, when set init MinimalGpMode, but set dynamic not MinimalGpMode

File: app/app_abci.go (hunk 2 above)
Comment (LeoGuo621) [reply app/app_abci.go]: done

File: app/app_abci.go (hunk 1 above)
Comment (LeoGuo621) [reply app/app_abci.go]: done

File: libs/cosmos-sdk/baseapp/baseapp_runtx.go [hunk 6]
```diff
@@ -305,6 +306,13 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx
 		return sdkerrors.ResponseDeliverTx(err, info.gInfo.GasWanted, info.gInfo.GasUsed, app.trace)
 	}
 
+	if app.updateGPOHandler != nil {
+		dgi := sdk.NewDynamicGasInfo(realTx.GetGasPrice(), info.gInfo.GasUsed)
+		dgis := make([]sdk.DynamicGasInfo, 1)
+		dgis[0] = dgi
+		app.updateGPOHandler(dgis)
```
Comment (zjg555543) [inline libs/cosmos-sdk/baseapp/baseapp_runtx.go]: Simplity the code like this?
app.updateGPOHandler([]sdk.DynamicGasInfo{sdk.NewDynamicGasInfo(realTx.GetGasPrice(), info.gInfo.GasUsed)})

File: libs/cosmos-sdk/baseapp/baseapp_runtx.go [hunk 7]
```diff
@@ -357,6 +365,14 @@ func (app *BaseApp) DeliverRealTx(txes abci.TxEssentials) abci.ResponseDeliverTx
 	if err != nil {
 		return sdkerrors.ResponseDeliverTx(err, info.gInfo.GasWanted, info.gInfo.GasUsed, app.trace)
 	}
+
+	if app.updateGPOHandler != nil {
+		dgi := sdk.NewDynamicGasInfo(realTx.GetGasPrice(), info.gInfo.GasUsed)
+		dgis := make([]sdk.DynamicGasInfo, 1)
+		dgis[0] = dgi
+		app.updateGPOHandler(dgis)
```
Comment (zjg555543) [inline libs/cosmos-sdk/baseapp/baseapp_runtx.go]: simplity code

File: libs/cosmos-sdk/baseapp/baseapp_runtx.go (hunk 6 above)
Comment (LeoGuo621) [reply libs/cosmos-sdk/baseapp/baseapp_runtx.go]: done

File: libs/cosmos-sdk/baseapp/baseapp_runtx.go (hunk 7 above)
Comment (LeoGuo621) [reply libs/cosmos-sdk/baseapp/baseapp_runtx.go]: done

File: app/app.go (hunk 5 above)
Comment (LeoGuo621) [reply app/app.go:708]: Fixed. When MinimalGpMode is set at initialization, the app also initializes the "app.gpo". And when DeliverTx or ParallelTxs, all modes will collect the gasPrice and gasUsed of tx.

File: app/app.go (hunk 4 above)
Comment (LeoGuo621) [reply app/app.go]: Fixed. When MinimalGpMode is set at initialization, the app also initializes the "app.gpo". And when DeliverTx or ParallelTxs, all modes will collect the gasPrice and gasUsed of tx.

File: app/types/gasprice.go (hunk 3 above)
Comment (LeoGuo621) [reply app/types/gasprice.go:13]: Yes, there are only 3 dynamic gp modes, but the NormalGpMode constant has not been used in the code implementation, and it is used here for indication.

**Review by zjg555543 — APPROVED** [review]

**Pull request merged**